src/shared/edit-util.c

```diff
@@ -18,19 +18,108 @@
 #include "strv.h"
 #include "tmpfile-util.h"
 
-void edit_file_free_all(EditFile **f) {
-        if (!f || !*f)
-                return;
+static void edit_file_context_done_internal(
+                EditFileContext *context,
+                bool clear) {
 
-        for (EditFile *i = *f; i->path; i++) {
-                free(i->path);
-                free(i->tmp);
+        int r;
+
+        assert(context);
+
+        FOREACH_ARRAY(i, context->files, context->n_files) {
+                if (i->temp) {
+                        (void) unlink(i->temp);
+                        free(i->temp);
+                }
+
+                if (context->remove_parent) {
+                        _cleanup_free_ char *parent = NULL;
+
+                        r = path_extract_directory(i->path, &parent);
+                        if (r < 0)
+                                log_debug_errno(r, "Failed to extract directory from '%s', ignoring: %m", i->path);
+
+                        /* No need to check if the dir is empty, rmdir does nothing if it is not the case. */
+                        (void) rmdir(parent);
+                }
+
+                if (clear)
+                        free(i->path);
+
+                free(i->original_path);
+                strv_free(i->comment_paths);
+        }
+
+        if (clear) {
+                context->files = mfree(context->files);
+                context->n_files = 0;
+        }
+}
+
+void edit_file_context_done(EditFileContext *context) {
+        return edit_file_context_done_internal(context, /* clear= */ false);
+}
+
+void edit_file_context_clear(EditFileContext *context) {
+        return edit_file_context_done_internal(context, /* clear= */ true);
+}
```

Comment:
Why we need these two?? Please free the array unconditionally, and always use `_done()` in the cleanup attribute, instead of `_clear()`